sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/Statistics.scala

```diff
@@ -58,60 +61,127 @@ case class Statistics(
-    Base64.encodeBase64String(statRow.asInstanceOf[UnsafeRow].getBytes)
+  /** Returns true iff the we support gathering column statistics on column of the given type. */
+  def supportsType(dataType: DataType): Boolean = dataType match {
+    case _: NumericType | TimestampType | DateType | BooleanType => true
+    case StringType | BinaryType => true
+    case _ => false
   }
-}
 
-object ColumnStat {
-  def apply(numFields: Int, str: String): ColumnStat = {
-    // use Base64 for decoding
-    val bytes = Base64.decodeBase64(str)
-    val unsafeRow = new UnsafeRow(numFields)
-    unsafeRow.pointTo(bytes, bytes.length)
-    ColumnStat(unsafeRow)
+  /**
+   * Creates a [[ColumnStat]] object from the given map. This is used to deserialize column stats
+   * from some external storage. The serialization side is defined in [[ColumnStat.toMap]].
+   */
+  def fromMap(dataType: DataType, map: Map[String, String]): Option[ColumnStat] = {
+    val str2val: (String => Any) = dataType match {
+      case _: IntegralType => _.toLong
+      case _: DecimalType => Decimal(_)
+      case DoubleType | FloatType => _.toDouble
+      case BooleanType => _.toBoolean
+      case _ => identity
+    }
+
+    try {
+      Some(ColumnStat(
+        ndv = map("ndv").toLong,
+        min = str2val(map.get("min").orNull),
+        max = str2val(map.get("max").orNull),
+        numNulls = map("numNulls").toLong,
+        avgLen = map.getOrElse("avgLen", "1").toLong,
+        maxLen = map.getOrElse("maxLen", "1").toLong
+      ))
+    } catch {
+      case NonFatal(e) =>
+        logWarning("Failed to parse column statistics", e)
+        None
+    }
   }
-}
 
-case class NumericColumnStat[T <: AtomicType](statRow: InternalRow, dataType: T) {
-  // The indices here must be consistent with `ColumnStatStruct.numericColumnStat`.
-  val numNulls: Long = statRow.getLong(0)
-  val max: T#InternalType = statRow.get(1, dataType).asInstanceOf[T#InternalType]
-  val min: T#InternalType = statRow.get(2, dataType).asInstanceOf[T#InternalType]
-  val ndv: Long = statRow.getLong(3)
-}
+  /**
+   * ndv: Long, min: T, max: T, numNulls: Long, avgLen: Long, maxLen: Long
+   */
+  def statExprs(col: Attribute, relativeSD: Double): CreateNamedStruct = {
+    def struct(exprs: Expression*): CreateNamedStruct = CreateStruct(exprs.map { expr =>
+      expr.transformUp { case af: AggregateFunction => af.toAggregateExpression() }
+    })
+    val zero = Literal(0, LongType)
+    val one = Literal(1, LongType)
 
-case class StringColumnStat(statRow: InternalRow) {
-  // The indices here must be consistent with `ColumnStatStruct.stringColumnStat`.
-  val numNulls: Long = statRow.getLong(0)
-  val avgColLen: Double = statRow.getDouble(1)
-  val maxColLen: Long = statRow.getInt(2)
-  val ndv: Long = statRow.getLong(3)
-}
+    // the approximate ndv should never be larger than the number of rows
+    val ndv = Least(Seq(HyperLogLogPlusPlus(col, relativeSD), Count(one)))
```

Comment:
I'm assuming `HyperLogLogPlusPlus` ignores like other aggregates. In that case shouldn't you use `Count(col)` instead of `Count(one)` as an upper bound ?